Client/Frontend/Browser/Toolbars/BottomToolbar/Menu/Bookmarks/AddEditBookmarkTableViewController.swift

```diff
@@ -0,0 +1,433 @@
+import Data
+import Shared
+
+private let log = Logger.browserLogger
+
+class AddEditBookmarkTableViewController: UITableViewController {
+    
+    private struct UX {
+        static let cellHeight: CGFloat = 44
+        /// Adds empty space between tableView's top and a custom header view.
+        static let headerTopInset = UIEdgeInsets(top: 36, left: 0, bottom: 0, right: 0)
+    }
+    
+    private enum DataSourcePresentationMode {
+        /// Showing currently selected save location.
+        case currentSelection
+        /// Showing a list of folders of which user can save the Bookmark to.
+        case folderHierarchy
+        
+        /// Flips between presentation modes.
+        mutating func toggle() {
+            switch self {
+            case .currentSelection: self = .folderHierarchy
+            case .folderHierarchy: self = .currentSelection
+            }
+        }
+    }
+    
+    // MARK: - View setup
+    
+    private lazy var saveButton: UIBarButtonItem = {
+        let button = UIBarButtonItem()
+        button.target = self
+        button.action = #selector(save)
+        button.title = Strings.SaveButtonTitle
+        
+        return button
+    }()
+    
+    private lazy var bookmarkDetailsView: BookmarkFormFieldsProtocol = {
+        switch mode {
+        case .addBookmark(let title, let url):
+            return BookmarkDetailsView(title: title, url: url)
+        case .addFolder(let title):
+            return FolderDetailsViewTableViewCell(title: title, viewHeight: UX.cellHeight)
+        case .editBookmark(let bookmark):
+            return BookmarkDetailsView(title: bookmark.displayTitle, url: bookmark.url)
+        case .editFolder(let folder):
+            return FolderDetailsViewTableViewCell(title: folder.displayTitle, viewHeight: UX.cellHeight)
+        }
+    }()
+    
+    // MARK: - Special cells
+    
+    /// A non-regular folder cell. Raw value represents a tag value of the cell
+    enum SpecialCell: Int {
+        case favorites = 10
+        case rootLevel = 11
+        case addFolder = 12
+    }
+    
+    let folderCellTag = 13
+    
+    /// Returns a count of how many non-folder cells should be visible(depends on Mode state)
+    private var specialButtonsCount: Int {
+        switch mode {
+        case .addFolder(_), .editFolder(_): return 1
+        case .addBookmark(_, _), .editBookmark(_): return 3
+        }
+    }
+    
+    private var rootLevelFolderCell: IndentedImageTableViewCell {
+        let cell = IndentedImageTableViewCell(image: #imageLiteral(resourceName: "menu_bookmarks")).then {
+            $0.folderName.text = Strings.BookmarkRootLevelCellTitle
+            $0.tag = SpecialCell.rootLevel.rawValue
+            if case .rootLevel = saveLocation, presentationMode == .folderHierarchy {
+                $0.accessoryType = .checkmark
+            }
+        }
+        
+        return cell
+    }
+    
+    private var favoritesCell: IndentedImageTableViewCell {
```

Comment:
This is really problematic, as it returns a new cell instance each time. Don't think you can use `then` or `let` constant here since it uses instance variables inside of it, but at least this should be `lazy` to avoid returning new instances.

Reply:
They need to be recreated when switching between presentation modes, otherwise it causes problems, the cell disappears when toggling between folder hierarchy/current selection